src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -0,0 +1,375 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.resource;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Maps;
+import com.mojang.logging.LogUtils;
+import net.minecraft.Util;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.PackType;
+import net.minecraftforge.common.ForgeConfig;
+import net.minecraftforge.common.ForgeConfigSpec;
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Cache manager for resources.
+ * <p>
+ * This class handles caching the resource listing calls on a pack, pack type and namespace level.
+ */
+public class ResourceCacheManager {
+
+    /**
+     * Indicates if the underlying namespaced managers need to support reloading.
+     * Disabled for the (in-jar / downloaded) default vanilla pack.
+     */
+    private final boolean supportsReloading;
+    /**
+     * Indicates if the indexing of the file-tree should happen off-thread or on-thread and block the process accordingly.
+     */
+    private final ForgeConfigSpec.BooleanValue indexOffThreadConfigOption;
+
+    /**
+     * The path builder (different users have different requirements for how they want to handle this)
+     */
+    private final BiFunction<PackType, String, Path> pathBuilder;
+    /**
+     * The individual sub managers by pack type and namespace
+     */
+    private final Map<PackTypeAndNamespace, NamespacedResourceCacheManager> managersByNamespace = new ConcurrentHashMap<>();
```

Comment:
And yes we need the atomicity, since we support off-thread indexing without holding up the main thread, so it is possible that both index hasCached are invoked simultaniously.